proc/exprswitch/exprswitch.go

```diff
@@ -1,123 +1,84 @@
 	"github.com/brimdata/zed/zbuf"
 )
 
 type ExprSwitch struct {
-	parent    proc.Interface
-	evaluator expr.Evaluator
+	*proc.Router
+	expr        expr.Evaluator
+	cases       map[string]*switchCase
+	defaultCase *switchCase
+}
 
-	cases     map[string]chan<- *zed.Value
-	defaultCh chan<- *zed.Value
-	doneChCh  chan chan<- *zed.Value
-	err       error
-	once      sync.Once
+var _ proc.Selector = (*ExprSwitch)(nil)
+
+type switchCase struct {
+	route proc.Interface
+	vals  []zed.Value
 }
 
-func New(parent proc.Interface, e expr.Evaluator) *ExprSwitch {
-	return &ExprSwitch{
-		parent:    parent,
-		evaluator: e,
-		cases:     make(map[string]chan<- *zed.Value),
-		doneChCh:  make(chan chan<- *zed.Value),
+func New(pctx *proc.Context, parent proc.Interface, e expr.Evaluator) *ExprSwitch {
+	router := proc.NewRouter(pctx, parent)
+	s := &ExprSwitch{
+		Router: router,
+		expr:   e,
+		cases:  make(map[string]*switchCase),
 	}
+	router.Link(s)
+	return s
 }
 
-func (s *ExprSwitch) NewProc(val *zed.Value) proc.Interface {
-	ch := make(chan *zed.Value)
+func (s *ExprSwitch) AddCase(val *zed.Value) proc.Interface {
+	route := s.Router.AddRoute()
 	if val == nil {
-		s.defaultCh = ch
+		s.defaultCase = &switchCase{route: route}
 	} else {
-		s.cases[string(val.Bytes)] = ch
+		s.cases[string(val.Bytes)] = &switchCase{route: route}
 	}
-	return &Proc{s, ch}
+	return route
 }
 
-func (s *ExprSwitch) run() {
-	defer func() {
-		for _, ch := range s.cases {
-			close(ch)
-		}
-		if s.defaultCh != nil {
-			close(s.defaultCh)
+func (s *ExprSwitch) Forward(router *proc.Router, batch zbuf.Batch) error {
+	ectx := batch.Context()
+	vals := batch.Values()
+	for i := range vals {
+		val := s.expr.Eval(ectx, &vals[i])
+		if val.IsMissing() {
+			continue
 		}
-		s.parent.Done()
-	}()
-	for {
-		batch, err := s.parent.Pull()
-		if batch == nil || err != nil {
-			s.err = err
-			return
+		which, ok := s.cases[string(val.Bytes)]
+		if !ok {
+			which = s.defaultCase
 		}
-		ectx := batch.Context()
-		vals := batch.Values()
-		for i := range vals {
-			val := s.evaluator.Eval(ectx, &vals[i])
-			if val.IsMissing() {
-				continue
-			}
-		again:
-			ch, ok := s.cases[string(val.Bytes)]
-			if !ok {
-				ch = s.defaultCh
-			}
-			if ch == nil {
-				continue
-			}
-			select {
-			case ch <- &vals[i]:
-			case doneCh := <-s.doneChCh:
-				s.handleDoneCh(doneCh)
-				if len(s.cases) == 0 && s.defaultCh == nil {
-					return
-				}
-				goto again
-			}
+		if which == nil {
+			continue
 		}
+		which.vals = append(which.vals, vals[i])
 	}
-}
-
-func (s *ExprSwitch) handleDoneCh(doneCh chan<- *zed.Value) {
-	if s.defaultCh == doneCh {
-		s.defaultCh = nil
-	} else {
-		for k, ch := range s.cases {
-			if ch == doneCh {
-				delete(s.cases, k)
-				break
-			}
+	// Send each case that has vals from this batch.
+	// We have vals that point into the current batch so we
+	// ref the batch for each outgoing new batch, then unref
+	// the batch once after the loop.
```

Comment:
This comment reads like this function calls Unref but it's actually the router that calls it.
```suggestion
	// We have vals that point into the current batch so we
	// ref the batch for each outgoing new batch.
```